### ISSUE TITLE
Search Functionality Issues with Bleeve Engine

### ISSUE THREAD

**Issue by amix307** [go-gitea/gitea#31565]
### Description

Hi,

I'm following up on my previous ticket: https://github.com/go-gitea/gitea/issues/30064. I always keep the service updated to the latest version, and the search problem persists even on version 1.22.1. I'm using the inbuilt Bleeve engine.

Here’s another example of the issue:
When searching for the string "services.gradle.org", there is one file in the repository where this string should be found. However, the Exact search method does not find the result, and the Fuzzy search hangs for about 2-3 minutes without finding anything.

To give more context about my instance: I have about 300 organizations and 3000 repositories, but overall the size is small as I don't have heavy files. I know for sure that there are many occurrences of the search string across default branches, likely several hundred. Despite this, the Exact search method returns only about 10 results and does so instantly. The Fuzzy search, however, hangs indefinitely. On version 1.21.11, I even encountered a 500 error and the Gitea service restarted.

Manual re-indexing does not help. I would like to resolve the search issues and have more transparent ways to understand how the code is indexed and to have more flexible control over Bleeve settings.

Thanks in advance for your help. I've attached some screenshots for reference.

### Gitea Version

1.22.1

### Can you reproduce the bug on the Gitea demo site?

No

### Screenshots

![2024-07-05_11-35-26](https://github.com/go-gitea/gitea/assets/24320787/c1cc9b9a-4f35-47af-8e9c-6b4276eb3ec2)
![2024-07-05_11-37-40](https://github.com/go-gitea/gitea/assets/24320787/3079344d-f429-49a0-9907-8134a78a60a4)
![2024-07-05_11-38-37](https://github.com/go-gitea/gitea/assets/24320787/50509b41-68ec-4a82-81d1-21a98abe76ca)
![2024-07-05_11-41-55](https://github.com/go-gitea/gitea/assets/24320787/6adc29af-7f4e-4842-b6de-d9830494ebfe)

### Git Version

2.31.1

### Operating System

CentOS Stream 9

### How are you running Gitea?

Self-Hosted from dl.gitea.org

### Database

PostgreSQL

**Comment by silverwind** [comment]
> the Fuzzy search hangs for about 2-3 minutes without finding anything.

I also noticed this search hanging starting with v1.22. If kept running, gitea continously consumes more and more memory up until the point where it exhausts the system resources. It's only for certain search terms, not all of them.

**Comment by amix307** [comment]
> > the Fuzzy search hangs for about 2-3 minutes without finding anything.
> 
> I also noticed this search hanging starting with v1.22. If kept running, gitea continously consumes more and more memory up until the point where it exhausts the system resources. It's only for certain search terms, not all of them.

Yes, yesterday my instance hangs and service gitea was restarted with OOM on VM 64gig ram)

**Comment by MICCustomsSolutions** [comment]
Same issue here. Eats all the RAM.
Disabling the index fixes the issue.

**Comment by carobme** [comment]
I can confirm this issue. Using the fuzzy code search with v1.22.1 (self-hosted) results in Gitea memory usage growing until the process runs out of memory.

Rebuilding the index (by stopping Gitea, `rm -rf indexers/*` and starting again) doesn't make a difference.

**Comment by techknowlogick** [comment]
Pinging @6543

**Comment by makar112233** [comment]
We're also had same issue

**Comment by smartEBL** [comment]
We are affected by this issue as well. A temporary fix for is changing
```ini
[indexer]
REPO_INDEXER_ENABLED = true
```
to 
```ini
[indexer]
REPO_INDEXER_ENABLED = false
```

Code search on single repository still works then, global search does not (of course). So it would be really nice to see that fixed. Is there anything we could provide for debugging?

**Comment by amix307** [comment]
@smartEBL now i switched to single-node elasticsearch in container at same vm and it works good, but have same issues with non transparent index mechanism (not searching all)

**Comment by nekdan** [comment]
We also encountered this issue and disabled the search functionality to ensure normal operation. However, this is a poor solution because we need the search functionality for our work.

**Comment by lunny** [comment]
Can anyone reproduce it in the development version? I think it's related to bleve versions.

**Comment by kemzeb** [comment]
> Can anyone reproduce it in the development version? I think it's related to bleve versions.

I am not yet familiar with how the code search indexing works nor am I familiar with bleve, but this could be the case after looking into the recent bleve releases.

In Gitea v1.21 we use bleve v2.30.10. Gitea v1.22 and onwards uses bleve [ v2.4.0](https://github.com/blevesearch/bleve/releases/tag/v2.4.0).

bleve [v2.4.1](https://github.com/blevesearch/bleve/releases/tag/v2.4.1) adds a fix to a memory leak problem associated to their "vector query path" (more specifically, it was a problem in a indirect dependency called [blevesearch/go-faiss](https://github.com/blevesearch/go-faiss/commit/693b06a12ca7c83f194973de009c8290de3ada90)).

I am not sure if this "path" is something that our code will eventually execute (or if it is only used by the vector indexing feature introduced in bleve v2.4.0), but I wish to bring this up to those that are maybe more familiar.

**Comment by silverwind** [comment]
According to go.mod, gitea v1.22 still uses bleve v2.3.10, so exact same version as v1.21:

https://github.com/go-gitea/gitea/blob/release/v1.22/go.mod#L22

**Comment by kemzeb** [comment]
After doing the following when firing up my dev Gitea instance:
- Create an empty repo
- Push Gitea's entire commit history to this empty repo (chose Gitea since its has a large history)
- Execute a fuzzy code search multiple times with the query "merge.go"

I was able to notice the huge memory cost. [Here](https://github.com/user-attachments/assets/83500580-a98d-41ea-97d1-f817fe0b7a07) is my generated pprof graph for reference.

I also checked out snapshot 1262ff6734543b37d834e63a6a623648c77ee4f4 (as this was before major changes were made to code search fuzzing) and I did not notice a performance impact on memory when observing my heap usage in the admin dashboard.

Don't have time to dig into this further yet, but thought this could be helpful in some way.

**Comment by Willy-J** [comment]
I am using the latest Docker version (1.22.1). After enabling the Repository Indexer and waiting for a few hours (I believe the indexing has completed as there isn't much code in the repository), I noticed that the number of zap files under `indexers/repos.bleve/store` changes almost every second, even though there has been no human activity during this period. I am not familiar with Bleve, so I am unsure if this is expected behavior.

**Comment by amix307** [comment]
> Bleve has been upgraded to v2.4.2 via #31762 but I don't whether this issue has been resolved.

i can try to test on nightly 1.22

**Comment by silverwind** [comment]
> @smartEBL now i switched to single-node elasticsearch in container at same vm and it works good, but have same issues with non transparent index mechanism (not searching all)

I switched to elasticsearch repo indexer as well now. It seems to be more efficient regarding disk usage than bleve as well, likely because of missing compression in bleve index.

**Comment by smartEBL** [comment]
@amix307 @silverwind 
How do your setups look like now? How did you change the app.ini to achieve that?

**Comment by amix307** [comment]
> @amix307 @silverwind How do your setups look like now? How did you change the app.ini to achieve that?

```
[indexer]

; before:

;ISSUE_INDEXER_NAME = gitea_issues
;ISSUE_INDEXER_TYPE = bleve
;MAX_FILE_SIZE = 1048576
;REPO_INDEXER_ENABLED = false
;REPO_INDEXER_EXCLUDE = resources/bin/**
;REPO_INDEXER_INCLUDE =
;REPO_INDEXER_PATH = indexers/repos.bleve
;REPO_INDEXER_TYPE = bleve
;STARTUP_TIMEOUT = 30s

; now:

REPO_INDEXER_ENABLED = true
REPO_INDEXER_REPO_TYPES = sources,forks,mirrors,templates
REPO_INDEXER_TYPE = elasticsearch
REPO_INDEXER_PATH = indexers/repos.bleve
REPO_INDEXER_CONN_STR = http://elastic:changeme@localhost:9200
REPO_INDEXER_NAME = gitea_codes
REPO_INDEXER_INCLUDE =
REPO_INDEXER_EXCLUDE =
MAX_FILE_SIZE = 1048576
```

**Comment by smartEBL** [comment]
> > @amix307 @silverwind How do your setups look like now? How did you change the app.ini to achieve that?
> 
> ```
> [indexer]
> 
> ; before:
> 
> ;ISSUE_INDEXER_NAME = gitea_issues
> ;ISSUE_INDEXER_TYPE = bleve
> ;MAX_FILE_SIZE = 1048576
> ;REPO_INDEXER_ENABLED = false
> ;REPO_INDEXER_EXCLUDE = resources/bin/**
> ;REPO_INDEXER_INCLUDE =
> ;REPO_INDEXER_PATH = indexers/repos.bleve
> ;REPO_INDEXER_TYPE = bleve
> ;STARTUP_TIMEOUT = 30s
> 
> ; now:
> 
> REPO_INDEXER_ENABLED = true
> REPO_INDEXER_REPO_TYPES = sources,forks,mirrors,templates
> REPO_INDEXER_TYPE = elasticsearch
> REPO_INDEXER_CONN_STR = http://elastic:changeme@localhost:9200
> REPO_INDEXER_NAME = gitea_codes
> REPO_INDEXER_INCLUDE =
> REPO_INDEXER_EXCLUDE =
> MAX_FILE_SIZE = 1048576
> ```

Cool, thanks for the fast reply. And how complicated is the setup/configuration of the elastic search? You also just spawn it inside some docker container? Or were there any big showstoppers you came across?

**Comment by silverwind** [comment]
We're getting offtopic, but it's pretty trivial, for example with `docker run`:

```bash
docker run --name="elasticsearch" \
  --volume="/opt/elasticsearch:/usr/share/elasticsearch/data" \
  --volume="/usr/share/zoneinfo:/usr/share/zoneinfo:ro" \
  --volume="/etc/localtime:/etc/localtime:ro" \
  --env="discovery.type=single-node" \
  --env="xpack.security.enabled=false" \
  --env="ingest.geoip.downloader.enabled=false" \
  --env="ELASTIC_USERNAME=elastic" \
  --env="ELASTIC_PASSWORD=changeme" \
  --publish="0.0.0.0:9200:9200" \
  --publish="0.0.0.0:9300:9300" \
  --restart="always" \
  --detach="true" \
  "elasticsearch:7.17.23"
```

**Comment by planbnet** [comment]
I just had this problem with a simple search on our local instance running Gitea 1.22.1. 
The installation has a few hundred repos and Bleve configured like this:
```
[indexer]
REPO_INDEXER_ENABLED = true
REPO_INDEXER_PATH = indexers/repos.bleve
MAX_FILE_SIZE = 1048576
REPO_INDEXER_INCLUDE =
REPO_INDEXER_EXCLUDE = client/bower_components/**,bower_components/**,node-modules/**,client/node-modules/**
```

16 GB of memory are slowly eaten up by the following goroutine (taken from Gitea's monitoring/stacktrace view). Clicking the trash can icon to kill the process does nothing.

`GET: /explore/code?q=build.yaml&fuzzy=true`

- `[...until here, there are always different function calls listed with each reload of the monitoring view,
   but the following stacktrace items will always be the same..]`

- `github.com/blevesearch/bleve/v2/search/searcher.(*PhraseSearcher).Next`
  `/go/pkg/mod/github.com/blevesearch/bleve/v2@v2.3.10/search/searcher/search_phrase.go:229`

- `github.com/blevesearch/bleve/v2/search/collector.(*TopNCollector).Collect`
  `/go/pkg/mod/github.com/blevesearch/bleve/v2@v2.3.10/search/collector/topn.go:228`

- `github.com/blevesearch/bleve/v2.(*indexImpl).SearchInContext`
  `/go/pkg/mod/github.com/blevesearch/bleve/v2@v2.3.10/index_impl.go:580`

- `code.gitea.io/gitea/modules/indexer/code/bleve.(*Indexer).Search`
  `/source/modules/indexer/code/bleve/bleve.go:285`

- `code.gitea.io/gitea/modules/indexer/code.PerformSearch`
  `/source/modules/indexer/code/search.go:138`

- `code.gitea.io/gitea/routers/web/explore.Code`
  `/source/routers/web/explore/code.go:80`

**Comment by johanvdw** [comment]
> Bleve has been upgraded to v2.4.2 via #31762 but I don't whether this issue has been resolved.

This is not solved on gitea 1.22.2

**Comment by planbnet** [comment]
Just tested it with 1.22.2, still the exact same issue.

To prevent unintended DOS attacks on our server, I've disabled the fuzzy search with a small nginx hack:

```
location /explore/code {                                                                                                                                                                                                                  
  if ($arg_fuzzy = 'true') {                                                                                                                                                                                                                
    rewrite ^/explore/code /explore/code?q=$arg_q&fuzzy=false last;                                                                                                                                                           
  }
  proxy_pass http://localhost:3000;                                                                                                                                                                                         }
  ```
  
If there's a cleaner workaround to keep the indexer, but disable fuzzy search, please let me know.

**Comment by jpraet** [comment]
I also got hit by this unpleasant surprise of OOM crashes after upgrading to gitea 1.22.

When such a significant problem is discovered in a new release, it would be nice if a warning would be added to the release notes and blog posts.

**Comment by jpraet** [comment]
Perhaps this one is related? https://github.com/go-gitea/gitea/pull/29706

**Comment by lunny** [comment]
> Perhaps this one is related? #29706

Did you mean it will happen only for fuzzy searching?

**Comment by jpraet** [comment]
Yes, exact search seems to be working fine, hence https://github.com/go-gitea/gitea/issues/31565#issuecomment-2333563036.

**Comment by johanvdw** [comment]
yes, it is a security issue.

**Comment by lunny** [comment]
> Just tested it with 1.22.2, still the exact same issue.
> 
> To prevent unintended DOS attacks on our server, I've disabled the fuzzy search with a small nginx hack:
> 
> ```
> location /explore/code {
>   if ($arg_fuzzy = 'true') {
>     rewrite ^/explore/code /explore/code?q=$arg_q&fuzzy=false last;
>   }
>   proxy_pass http://localhost:3000;
> }
> ```
> 
> If there's a cleaner workaround to keep the indexer, but disable fuzzy search, please let me know.

Maybe we can have an option to disable fuzzy search at first.

**Comment by wxiaoguang** [comment]
I can see some related bleeve related fuzzy changes in 1.22 : Determine fuzziness of bleve indexer by keyword length #29706  , Expose fuzzy search for issues/pulls #29701 


Just a guess, would revert #29706 resolve the problem? @6543

**Comment by 6543** [comment]
:eyes:

**Comment by lunny** [comment]
We can temporarily disable fuzzy search for the bleve search engine in v1.23.

**Comment by wxiaoguang** [comment]
> I can see some related bleeve related fuzzy changes in 1.22 : Determine fuzziness of bleve indexer by keyword length #29706 , Expose fuzzy search for issues/pulls #29701
> 
> Just a guess, would reverting #29706 resolve the problem? also cc @6543

@6543 then we need another dirty patch to the indexer system:  Fix bleve fuzziness search #33078
